addons/mail/static/src/messaging/entity/entity/entity.js

```diff
@@ -0,0 +1,824 @@
+         */
+        static get all() {
+            const { state } = this.env.store;
+            return Object.values(state.entities).filter(entity => entity instanceof this);
+        }
+
+        /**
+         * Single way to create new entity instances in the store.
+         *
+         * @static
+         * @param {Object} data data object with initial data, including relations.
+         * @throws {Error} in case Entity class is a singleton and it creates
+         *   more than one instance.
+         * @returns {mail.messaging.entity.Entity} newly created entity
+         */
+        static create(data) {
+            const { state } = this.env.store;
+            if (this.isSingleton && this.all.length > 0) {
+                throw new Error(`Singleton Entity class "${this.name}" cannot create more than one instance.`);
+            }
+            const entity = new this(data);
+            state.entities[entity.localId] = entity;
+            // ensure observable version of entity is handled.
+            const proxifiedEntity = state.entities[entity.localId];
+            this.__update(proxifiedEntity, data);
+            return proxifiedEntity;
+        }
+
+        /**
+         * @static
+         */
+        static delete() {
+            for (const entity of this.all) {
+                entity.delete();
+            }
+            this.__delete(this);
+        }
+
+        /**
+         * @static
+         * @param {any} id
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static fromId(id) {
+            const allEntities = this.all;
+            return allEntities.find(entity => entity.id === id);
+        }
+
+        /**
+         * @static
+         * @param {string|mail.messaging.entity.Entity|undefined} entityOrLocalId
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static get(entityOrLocalId) {
+            const { state } = this.env.store;
+            if (entityOrLocalId === undefined) {
+                return undefined;
+            }
+            const entity = state.entities[
+                entityOrLocalId.isEntity
+                    ? entityOrLocalId.localId
+                    : entityOrLocalId
+            ];
+            if (!(entity instanceof this) && entity !== this) {
+                return;
+            }
+            return entity;
+        }
+
+        static link(data) {
+            this.__link(this, data);
+        }
+
+        /**
+         * @static
+         * @param {Object} data
+         * @param {any} data.id
+         * @returns {mail.messaging.entity.Entity}
+         */
+        static insert(data) {
+            let entity = this.fromId(data.id);
+            if (!entity) {
+                entity = this.create(data);
+            } else {
+                entity.update(data);
+            }
+            return entity;
+        }
+
+        /**
+         * @throws {Error} when called with non singleton classes
```

Comment:
I don't like singleton and I don't think it is necessary, especially at the entity core level. There is no reason to arbitrary limit an entity to one.